src/Aspire.Hosting.Docker/DockerComposePublishingContext.cs

```diff
@@ -0,0 +1,521 @@
+                        IsHttpOnly = g.All(e => e.UriScheme is "http" or "https"),
+                        AnyH2 = g.Any(e => e.Transport is "http2"),
+                        UniqueSchemes = g.Select(e => e.UriScheme).Distinct().ToArray(),
+                        Index = g.Min(e => endpointIndexMap[e.Name]),
+                    })
+                .ToList();
+
+            foreach (var endpointsByPort in endpointsByTargetPort)
+            {
+                foreach (var e in endpointsByPort.Endpoints)
+                {
+                    var port = e.Port.GetValueOrDefault();
+                    var targetPort = e.TargetPort ?? e.Port.GetValueOrDefault();
+
+                    _endpointMapping[e.Name] = new(e.UriScheme, resource.Name, port, targetPort, false, e.IsExternal);
+                }
+            }
+
+            return;
+
+            static bool IsHttpScheme(string scheme) => scheme is "http" or "https";
+        }
+
+        private async Task ProcessArgumentsAsync(CancellationToken cancellationToken)
+        {
+            if (resource.TryGetAnnotationsOfType<CommandLineArgsCallbackAnnotation>(out var commandLineArgsCallbackAnnotations))
+            {
+                var context = new CommandLineArgsCallbackContext([], cancellationToken: cancellationToken);
+
+                foreach (var c in commandLineArgsCallbackAnnotations)
+                {
+                    await c.Callback(context).ConfigureAwait(false);
+                }
+
+                foreach (var arg in context.Args)
+                {
+                    var value = await ProcessValueAsync(arg).ConfigureAwait(false);
+
+                    if (value is not string str)
+                    {
+                        throw new NotSupportedException("Command line args must be strings");
+                    }
+
+                    Args.Add(new(str));
+                }
+            }
+        }
+
+        private async Task ProcessEnvironmentAsync(DistributedApplicationExecutionContext executionContext, CancellationToken cancellationToken)
+        {
+            if (resource.TryGetAnnotationsOfType<EnvironmentCallbackAnnotation>(out var environmentCallbacks))
+            {
+                var context = new EnvironmentCallbackContext(executionContext, cancellationToken: cancellationToken);
+
+                foreach (var c in environmentCallbacks)
+                {
+                    await c.Callback(context).ConfigureAwait(false);
+                }
+
+                foreach (var kv in context.EnvironmentVariables)
+                {
+                    var value = await ProcessValueAsync(kv.Value).ConfigureAwait(false);
+
+                    EnvironmentVariables.Add(new ComposeEnvironmentVariable(kv.Key, value.ToString()));
+                }
+
+                if (resource is ProjectResource)
+                {
+                    SetUpAspNetCoreUrlsForProject();
+                }
```

Comment:
Why do you need this? This should be part of the project resource.

Reply:
Agreed. Will remove. Port allocator just needs to track target ports on the host not the container ports